sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java

```diff
@@ -0,0 +1,387 @@
+    if (segments.isEmpty()) {
+      return null;
+    }
+    String field = segments.remove(0);
+    Value value = valueMap.get(field);
+    if (segments.isEmpty()) {
+      return value;
+    }
+    // Field path traversal is not done, recurse into map values.
+    if (value == null || !value.getValueTypeCase().equals(ValueTypeCase.MAP_VALUE)) {
+      return null;
+    }
+    return findMapValue(segments, value.getMapValue().getFieldsMap());
+  }
+
+  private static class OrderByFieldPath implements Comparable<OrderByFieldPath> {
+
+    public static OrderByFieldPath fromString(String fieldPath) {
+      if (fieldPath.isEmpty()) {
+        throw new IllegalArgumentException("Could not resolve empty field path");
+      }
+      List<String> segments = new ArrayList<>();
+      while (!fieldPath.isEmpty()) {
+        Matcher segmentMatcher = FIELD_PATH_SEGMENT_REGEX.matcher(fieldPath);
+        boolean foundMatch = segmentMatcher.lookingAt();
+        if (!foundMatch) {
+          throw new IllegalArgumentException("OrderBy field path was malformed");
+        }
+        String fieldName;
+        if ((fieldName = segmentMatcher.group(1)) != null) {
+          segments.add(fieldName);
+        } else if ((fieldName = segmentMatcher.group(2)) != null) {
+          String unescaped = unescapeFieldName(fieldName.substring(1, fieldName.length() - 1));
+          segments.add(unescaped);
+        } else {
+          throw new IllegalArgumentException("OrderBy field path was malformed");
+        }
+        fieldPath = fieldPath.substring(fieldName.length());
+        // Due to the regex, any non-empty fieldPath will have a dot before the next nested field.
+        if (fieldPath.startsWith(".")) {
+          fieldPath = fieldPath.substring(1);
+        }
+      }
+      return new OrderByFieldPath(ImmutableList.copyOf(segments));
+    }
+
+    private final ImmutableList<String> segments;
+
+    private OrderByFieldPath(ImmutableList<String> segments) {
+      this.segments = segments;
+    }
+
+    public String canonicalString() {
+      return String.join(".", segments);
+    }
+
+    public boolean isDocumentName() {
+      return segments.size() == 1 && "__name__".equals(segments.get(0));
+    }
+
+    public ImmutableList<String> getSegments() {
+      return segments;
+    }
+
+    @Override
+    public boolean equals(@Nullable Object other) {
+      if (other instanceof OrderByFieldPath) {
+        return this.segments.equals(((OrderByFieldPath) other).getSegments());
+      }
+      return super.equals(other);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(segments);
+    }
+
+    @Override
+    public int compareTo(OrderByFieldPath other) {
+      // Taken from com.google.cloud.firestore.FieldPath.
+      int length = Math.min(this.getSegments().size(), other.getSegments().size());
+      for (int i = 0; i < length; i++) {
+        int cmp = this.getSegments().get(i).compareTo(other.getSegments().get(i));
```

Comment:
UTF16 does not follow unicode code point ordering. Can we use something like the ICU4J library which has better support for unicode string comparison here? Or compare the UTF8 bytes.

Reply:
So I guess the Firestore SDK FieldPath implementation is wrong? I've switched it to comparing UTF8 bytes.